### ISSUE TITLE
Failed to build with c++23

### ISSUE THREAD

**Issue by DezlowNG** [KhronosGroup/KTX-Software#656]
Hello!

I'm always get error: ```non-const lvalue reference to type 'cppspmd_sse41::spmd_kernel::vfloat' cannot bind to a temporary of type 'cppspmd_sse41::spmd_kernel::vfloat``` when I'm trying to build KTX.

```
In file included from /media/dezlow/Drive/Dev/C++/Oneiro/ThirdParty/KTX/lib/basisu/encoder/basisu_kernels_sse.cpp:41:
/media/dezlow/Drive/Dev/C++/Oneiro/ThirdParty/KTX/lib/basisu/encoder/cppspmd_sse.h:496:10: error: non-const lvalue reference to type 'cppspmd_sse41::spmd_kernel::vfloat' cannot bind to a temporary of type 'cppspmd_sse41::spmd_kernel::vfloat'
                return dst;
                       ^~~
/media/dezlow/Drive/Dev/C++/Oneiro/ThirdParty/KTX/lib/basisu/encoder/cppspmd_sse.h:508:10: error: non-const lvalue reference to type 'cppspmd_sse41::spmd_kernel::vfloat' cannot bind to a temporary of type 'cppspmd_sse41::spmd_kernel::vfloat'
                return dst;
                       ^~~
```

I'm building my project with using GitHub Actions and I get the same error.

Sorry for my English, lol.

Thanks!

**cat /etc/os-release**
```
PRETTY_NAME="Debian GNU/Linux bookworm/sid"
NAME="Debian GNU/Linux"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

**clang --version**
```
Debian clang version 14.0.6-2
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
```

**Comment by MarkCallow** [comment]
Our CI builds with clang for macOS, Windows ARM and Emscripten. In addition macOS/Xcode 14 (clang 14.0.29) is my primary development environment. CI also builds on Ubuntu with GCC. So the issue is something specific to your situation therefore not easy for me to debug. The error appears to be related to the use of `&&` on the `dst` parameter declaration but why only you are seeing it I have no idea. Please try to debug further yourself.

**Comment by MarkCallow** [comment]
Changed title because KTX builds fine on Linux with GCC.

**Comment by MarkCallow** [comment]
Is this still an issue for you in `main`?

**Comment by Flone-dnb** [comment]
I just got the same error on Archlinux with Clang 16.

**Comment by MarkCallow** [comment]
> I just got the same error on Archlinux with Clang 16.

On which branch? Which errors?

The latest code compiles via ClangCL both 15 & 16 in VS 2022. It compiles with clang 14 on macOS and clang 17 in Emscripten. I do not have the bandwidth to work on another platform. If you want this fixed I'll need a PR that fixes it. If the errors are in the `lib/basisu/*` code then the preferred fix is to disable the subject warnings in `CMakeLists.txt`. If elsewhere in `lib` then please fix the warnings.

**Comment by Flone-dnb** [comment]
> On which branch? Which errors?

Latest commit from `main` branch. Same errors as described above.

Changing 2 places in `cppspmd_sse.h` to:

> return static_cast<vfloat&>(dst)

Fixes build.

**Comment by MarkCallow** [comment]
I can't debug this or test fixes as my machine is arm64-based so SSE code is never compiled. I made a simple reproducer which I'm compiling with AppleClang 15.0.0. The error appears when compiling with `--std=c++2b` (i.e. draft c++23).

@Flone-dnb and @DezlowNG are you building with KTX-Software's CMakeLists.txt? Do you have a parent project that is including it? I'm guessing you must somehow be specifying C++23. Looks like I'll need to explicitly specify the c++ standard when compiling the library. First I'll see if I can find a code fix.

**Comment by MarkCallow** [comment]
I opened an issue upstream: https://github.com/BinomialLLC/basis_universal/issues/366.

**Comment by MarkCallow** [comment]
Fixed by PR #806.

**Comment by Flone-dnb** [comment]
@MarkCallow 

> are you building with KTX-Software's CMakeLists.txt? Do you have a parent project that is including it? I'm guessing you must somehow be specifying C++23

That's right.

I just tried using the latest commit (f8220ec) and I can confirm that everything builds correctly on Linux now.

Thanks.
